iothub/device/src/IotHubDeviceClient.cs

```diff
@@ -191,175 +106,89 @@ public Task SendEventBatchAsync(IEnumerable<Message> messages, CancellationToken
-        public Task SetMethodDefaultHandlerAsync(
-            Func<DirectMethodRequest, object, Task<DirectMethodResponse>> methodHandler,
-            object userContext,
-            CancellationToken cancellationToken = default)
-            => InternalClient.SetMethodDefaultHandlerAsync(methodHandler, userContext, cancellationToken);
+            // Wait to acquire the _deviceReceiveMessageSemaphore. This ensures that concurrently invoked
+            // SetReceiveMessageHandlerAsync calls are invoked in a thread-safe manner.
+            await _deviceReceiveMessageSemaphore.WaitAsync(cancellationToken).ConfigureAwait(false);
 
-        /// <summary>
-        /// Retrieve the device twin properties for the current device. DeviceClient instance must be opened already.
-        /// For the complete device twin object, use Microsoft.Azure.Devices.RegistryManager.GetTwinAsync(string deviceId).
-        /// </summary>
-        /// <param name="cancellationToken">A cancellation token to cancel the operation.</param>
-        /// <exception cref="InvalidOperationException">Thrown if DeviceClient instance is not opened already.</exception>
-        /// <exception cref="IotHubClientException">Thrown and <see cref="IotHubClientException.StatusCode"/> is set to <see cref="IotHubStatusCode.NetworkErrors"/>
-        /// when the operation has been canceled. The inner exception will be <see cref="OperationCanceledException"/>.</exception>
-        /// <returns>The device twin object for the current device</returns>
-        public Task<Twin> GetTwinAsync(CancellationToken cancellationToken = default) => InternalClient.GetTwinAsync(cancellationToken);
+            try
+            {
+                // If a ReceiveMessageCallback is already set on the DeviceClient, calling SetReceiveMessageHandlerAsync
+                // again will cause the delegate to be overwritten.
+                if (messageHandler != null)
+                {
+                    // If this is the first time the delegate is being registered, then the telemetry downlink will be enabled.
+                    await EnableReceiveMessageAsync(cancellationToken).ConfigureAwait(false);
+                    _deviceReceiveMessageCallback = new Tuple<Func<Message, object, Task>, object>(messageHandler, userContext);
+                }
+                else
+                {
+                    // If a null delegate is passed, it will disable the callback triggered on receiving messages from the service.
+                    _deviceReceiveMessageCallback = null;
+                    await DisableReceiveMessageAsync(cancellationToken).ConfigureAwait(false);
+                }
+            }
+            catch (IotHubClientException ex) when (ex.InnerException is OperationCanceledException)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+                throw;
+            }
```

Comment:
Can we go upstream to find what is converting an OperationCanceledException into an IotHubClientException and not do that?

Reply:
I'm not seeing an OperationCancelledException being converted to an IotHubException in the amqp flow. This is possibly leftover code from our previous exception handling logic. Since we're rewriting the mqtt flow anyway, I'll go ahead and remove this block.

Reply:
Actually, scratch that, this block will do the conversion in our delegating handler pipeline: 
https://github.com/Azure/azure-iot-sdk-csharp/blob/f512c192d2b5b9f55a7e4a7c433f4969cd0e96a4/iothub/device/src/Pipeline/ErrorDelegatingHandler.cs#L202-L205
https://github.com/Azure/azure-iot-sdk-csharp/blob/f512c192d2b5b9f55a7e4a7c433f4969cd0e96a4/iothub/device/src/Pipeline/ErrorDelegatingHandler.cs#L29-L38

We should remove OperationCanceledException from the list of expected network exceptions.

Reply:
Are we going to add a task for that?

Reply:
Syncing up with Bryce to get this added to a task.